synapse/federation/federation_client.py

```diff
@@ -1113,14 +1114,17 @@ async def get_public_rooms(
                 requests over federation
 
         """
-        return await self.transport_layer.get_public_rooms(
-            remote_server,
-            limit,
-            since_token,
-            search_filter,
-            include_all_networks=include_all_networks,
-            third_party_instance_id=third_party_instance_id,
-        )
+        try:
+            return await self.transport_layer.get_public_rooms(
+                remote_server,
+                limit,
+                since_token,
+                search_filter,
+                include_all_networks=include_all_networks,
+                third_party_instance_id=third_party_instance_id,
+            )
+        except (RequestSendFailed, HttpResponseException):
+            raise SynapseError(502, "Failed to fetch room list")
```

Comment:
I don't think we should be picking error codes here: that should be handled in the REST server code, not the federation client code.

Reply:
Yeees, though what are you envisaging exactly? We can add an exception type for `CouldNotTalkToRemote`, but would that extend `SynapseError` (which is what we usually do), or handled explicitly at the rest layer, or the servlet layer?

Or would it be easier to hoist this back up into the room list handler again?

Reply:
Right. I notice that `get_remote_public_room_list` has yet another layer of exception handling: https://github.com/matrix-org/synapse/blob/4f65b5d2d4c8f09d46547e8d367aa16ac51008c7/synapse/handlers/room_list.py#L378-L390. Indeed, it looks like #10400 might have messed this up.

And, if that weren't enough of a mess, `PublicRoomListRestServlet` has *another* round of catching `HttpResponseException`, which I *think* is meant to be unreachable, but who the hell knows.

I think you should allow the `RequestSendFailed` / `HttpResponseException`s to propagate into `RoomListHandler` and raise an appropriate SynapseError there; and remove the handling from `PublicRoomListRestServlet`. And, for the love of $deity, document the behaviour of these things.

Reply:
Wheeeeeee untested features